publisher/publisher.go

```diff
@@ -0,0 +1,292 @@
+	"time"
+
+	"github.com/letsencrypt/boulder/core"
+	blog "github.com/letsencrypt/boulder/log"
+)
+
+type LogDescription struct {
+	ID        []byte
+	URI       string
+	PublicKey *ecdsa.PublicKey
+}
+
+type rawLogDescription struct {
+	URI       string `json:"uri"`
+	PublicKey string `json:"key"`
+}
+
+func (logDesc *LogDescription) UnmarshalJSON(data []byte) error {
+	var rawLogDesc rawLogDescription
+	if err := json.Unmarshal(data, &rawLogDesc); err != nil {
+		return fmt.Errorf("Failed to unmarshal log description, %s", err)
+	}
+	logDesc.URI = rawLogDesc.URI
+	// Load Key
+	pkBytes, err := base64.StdEncoding.DecodeString(rawLogDesc.PublicKey)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+	pk, err := x509.ParsePKIXPublicKey(pkBytes)
+	if err != nil {
+		return fmt.Errorf("")
+	}
+	switch k := pk.(type) {
+	case ecdsa.PublicKey:
+		logDesc.PublicKey = &k
+	case *ecdsa.PublicKey:
+		logDesc.PublicKey = k
+	default:
+		return fmt.Errorf("Failed to unmarshal log description for %s, unsupported public key type", logDesc.URI)
+	}
+
+	// Generate key hash for log ID
+	pkHash := sha256.Sum256(pkBytes)
+	logDesc.ID = pkHash[:]
+	if len(logDesc.ID) != 32 {
+		return fmt.Errorf("Invalid log ID length [%d]", len(logDesc.ID))
+	}
+
+	return nil
+}
+
+// CTConfig defines the JSON configuration file schema
+type CTConfig struct {
+	Logs              []LogDescription `json:"logs"`
+	SubmissionRetries int              `json:"submissionRetries"`
+	// This should use the same method as the DNS resolver
+	SubmissionBackoffString string        `json:"submissionBackoff"`
+	SubmissionBackoff       time.Duration `json:"-"`
+
+	BundleFilename string   `json:"intermediateBundleFilename"`
+	IssuerBundle   []string `json:"-"`
+}
+
+type ctSubmissionRequest struct {
+	Chain []string `json:"chain"`
+}
+
+const (
+	sctVersion       = 0
+	sctSigType       = 0
+	sctX509EntryType = 0
+)
+
+// PublisherAuthorityImpl defines a Publisher
+type PublisherAuthorityImpl struct {
+	log *blog.AuditLogger
+	CT  *CTConfig
+	SA  core.StorageAuthority
+}
+
+// NewPublisherAuthorityImpl creates a Publisher that will submit certificates
+// to any CT logs configured in CTConfig
+func NewPublisherAuthorityImpl(ctConfig *CTConfig) (PublisherAuthorityImpl, error) {
+	var pub PublisherAuthorityImpl
+
+	logger := blog.GetAuditLogger()
+	logger.Notice("Publisher Authority Starting")
+	pub.log = logger
+
+	if ctConfig != nil {
+		pub.CT = ctConfig
+		if ctConfig.BundleFilename != "" {
+			bundle, err := core.LoadCertBundle(ctConfig.BundleFilename)
+			if err != nil {
+				return pub, nil
+			}
+			for _, cert := range bundle {
+				pub.CT.IssuerBundle = append(pub.CT.IssuerBundle, base64.StdEncoding.EncodeToString(cert.Raw))
+			}
+		}
+		ctBackoff, err := time.ParseDuration(ctConfig.SubmissionBackoffString)
+		if err != nil {
+			return pub, err
+		}
+		pub.CT.SubmissionBackoff = ctBackoff
+	}
+
+	return pub, nil
+}
+
+// SubmitToCT will submit the certificate represented by certDER to any CT
```

Comment:
Comment doesn't match function signature, which takes an x509.Certificate. Probably better to take a DER, as the comment says, in case the round trip into an x509.Certificate and back is lossy (shouldn't be, but just in case).